storage/innobase/xtrabackup/src/ds_decompress_zstd.cc

```diff
@@ -0,0 +1,341 @@
+  block_info.last_block = (cont_header)&LAST_BLOCK_MASK;
+  block_info.block_type =
+      static_cast<block_type_t>((cont_header >> 1) & BLOCK_TYPE_MASK);
+  block_info.data_size = (cont_header >> 3);
+
+  if (block_info.block_type == ZSTD_BLOCK_RAW ||
+      block_info.block_type == ZSTD_BLOCK_COMPRESSED) {
+    block_info.data = stream.ptr(block_info.data_size);
+    /* empty file will have data_size = 0  */
+    if (block_info.data_size > 0 && block_info.data == nullptr) {
+      return ZSTD_INCOMPLETE;
+    }
+    return ZSTD_OK;
+  } else if (block_info.block_type == ZSTD_BLOCK_RLE) {
+    /* 1 byte repeatable block */
+    block_info.data = stream.ptr(1);
+    if (block_info.data == nullptr) {
+      return ZSTD_INCOMPLETE;
+    }
+    return ZSTD_OK;
+  }
+  return ZSTD_ERROR;
+}
+
+/* check if we have a full frame to decode */
+ZSTD_stream::err_t ZSTD_stream::next_frame(size_t &frame_size) {
+  if (state != ZSTD_STREAM_STATE_BEFORE_FRAME_HEADER) return ZSTD_ERROR;
+
+  size_t start_frame_cur, end_frame_cur, start_frame_pos, end_frame_pos = 0;
+  if (state == ZSTD_STREAM_STATE_BEFORE_FRAME_HEADER) {
+    /* parse frame header */
+    stream.save_pos();
+    start_frame_cur = stream.get_saved_cur();
+    start_frame_pos = stream.get_saved_pos();
+    const auto err = parse_frame_header();
+    if (err == ZSTD_ERROR) {
+      /* parse error, exit with error */
+      return ZSTD_ERROR;
+    } else if (err == ZSTD_INCOMPLETE) {
+      /* incomplete frame header, save remaining data and exit normally */
+      stream.restore_pos();
+      return ZSTD_INCOMPLETE;
+    }
+    state = ZSTD_STREAM_STATE_BEFORE_BLOCK_HEADER;
+  }
+  if (state == ZSTD_STREAM_STATE_BEFORE_BLOCK_HEADER) {
+    block_info_t block_info;
+    do {
+      /* parse block */
+      const auto err = parse_block(block_info);
+      if (err == ZSTD_ERROR) {
+        /* parse error */
+        return ZSTD_ERROR;
+      } else if (err == ZSTD_INCOMPLETE) {
+        /* incomplete block, save remaining data and exit normally */
+        stream.restore_pos();
+        state = ZSTD_STREAM_STATE_BEFORE_FRAME_HEADER;
+        return ZSTD_INCOMPLETE;
+      }
+      if (block_info.last_block) {
+        state = ZSTD_STREAM_STATE_BEFORE_FRAME_HEADER;
+        if (frame_header.checksumFlag) {
+          if (!stream.read_u32_le(&stored_checksum)) {
+            /* incomplete block, save remaining data and exit normally */
+            stream.restore_pos();
+            return ZSTD_INCOMPLETE;
+          }
+        }
+        end_frame_cur = stream.get_current_cur();
+        end_frame_pos = stream.get_current_pos();
+        frame_size = stream.length_from_to(start_frame_cur, start_frame_pos,
+                                           end_frame_cur, end_frame_pos);
+        stream.restore_pos(start_frame_cur, start_frame_pos);
+        /* we have a complete frame */
+        return ZSTD_OK;
+      }
+    } while (!block_info.last_block);
+  }
+  return ZSTD_ERROR;
+}
```

Comment:
You shouldn't read ZSTD headers by yourself. `ZSTD_findFrameCompressedSize` should allow to check if we have a full frame:
```
size_t ZSTD_findFrameCompressedSize(const void* src, size_t srcSize);
 `src` should point to the start of a ZSTD frame or skippable frame.
 `srcSize` must be >= first frame size
 @return : the compressed size of the first frame starting at `src`,
           suitable to pass as `srcSize` to `ZSTD_decompress` or similar,
        or an error code if input is invalid 
```

Reply:
@inikep I'm using `ZSTD_getFrameHeader` to parse the frame header. I just have a custom function `parse_frame_header` so I can return `ZSTD_stream::err_t` to `ZSTD_stream::next_frame` which has the logic to concatenate the buffers in case of partial frame has been received.

Reply:
@altmannmarcelo I think you can reimplement `ZSTD_stream::next_frame` using just `ZSTD_findFrameCompressedSize` and get rid of ZSTD header parsing because you don't need to go to such a low level. You should assume that working with ZSTD headers is prohibited because ZSTD API functions should be enough for our needs.

Reply:
re-wrote `ZSTD_stream::next_frame` using `ZSTD_findFrameCompressedSize` - it's way simpler now!